dali/benchmark/displacement_cpu_bench.cc

```diff
@@ -0,0 +1,122 @@
+#include "dali/pipeline/pipeline.h"
+#include "dali/util/image.h"
+#include "dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h"
+#include "dali/pipeline/operators/displacement/flip.h"
+#include "dali/pipeline/operators/displacement/rotate.h"
+#include "dali/pipeline/operators/displacement/sphere.h"
+#include "dali/pipeline/operators/displacement/water.h"
+#include "dali/pipeline/data/tensor.h"
+
+namespace dali {
+
+namespace {
+template <typename OpType>
+struct OpInfo {
+  static OpSpec op;
+};
+template <typename OpType>
+OpSpec OpInfo<OpType>::op = OpSpec("OperatorName");
+}
+
+#define DALI_DECLARE_OP_INFO(OP_TYPE, ...) \
+namespace {                                \
+template<>                                 \
+OpSpec OpInfo<OP_TYPE>::op = __VA_ARGS__;  \
+}
+
+// DisplacementFilterType expects type with CPUBackend, for example DisplacementFilter<CPUBackend>
+// T is underlying input data
+template <typename DisplacementFilterType, typename T>
+void DisplacementBench(benchmark::State& st) {//NOLINT
+  OpSpec op = OpInfo<DisplacementFilterType>::op;
+  int interp_type = st.range(0);
+  op.AddArg("interp_type", interp_type);
+
+  // batch_size and num_threads are checked by OperatorBase, and are used later to create
+  // appropriate executor by the Pipeline. We have to specify them,
+  // but they are not used in this benchmark
+  const int dummy_batch_size = 128;
+  const int dummy_num_thread = 1;
+  op.AddArg("batch_size", dummy_batch_size).AddArg("num_threads", dummy_num_thread);
+
+  // Create Operator and pass its arguments via OpSpec
+  // We get partially-specified OpSpec, added remaining arguments,
+  // and we use it with DisplacementFilterType to instantiate operator
+  DisplacementFilterType df(op);
+
+  static constexpr int W = 1920, H = 1080, C = 3;
+  static constexpr int N = W * H * C;
+
+  // The inputs and outputs to CPUBackend are: shared_ptr<Tensor<CPUBackend>>;
+  // create input and output, initialize input
+  shared_ptr<Tensor<CPUBackend>> tensor_in(new Tensor<CPUBackend>());
+  shared_ptr<Tensor<CPUBackend>> tensor_out(new Tensor<CPUBackend>());
+  // If we want to specify input, we can share data
+  // tensor_in->ShareData(img, N * sizeof(T));
+  // Here we let underlying buffer allocate it by itself. We have to specify size and type
+  tensor_in->set_type(TypeInfo::Create<T>());
+  tensor_in->Resize({W, H, C});
+  // tensor out is resized by operator itself in DisplacementFilter::DataDependentSetup()
+
+  // TODO(klecki) Accomodate to use different inputs from test data
+  auto *ptr = tensor_in->mutable_data<T>();
+  for (int i = 0; i < N; i++) {
+    ptr[i] = i;
+  }
+
+  // Create workspace and set input and output
+  SampleWorkspace s_ws;
+  s_ws.AddInput(tensor_in);
+  s_ws.AddOutput(tensor_out);
+
+  // Run once so output is allocated
+  df.Run(&s_ws);
+
+  for (auto _ : st) {
+    df.Run(&s_ws);
+  }
+}
+
+#define DALI_BENCHMARK_DISPLACEMENT(OP_TYPE)            \
```

Comment:
Could this code be made with templates instead of macros? I could imagine, that somebody needs to take this code are write his own benchmark. He will have something similar to what we had with unit tests...